### PR TITLE
Add freestyle notification for deployment

### PR DESCRIPTION
**Pull request by khughes-atlassian**
Create a post-deployment notifier to send deployment information to Bitbucket Server.

Opening up the PR for an initial review. Still todo: Unit tests

Screenshots:
Screen Shot 2021-07-20 at 12 52 24 pm
Screen Shot 2021-07-20 at 5 46 19 pm
Screen Shot 2021-07-20 at 5 46 39 pm
Screen Shot 2021-07-20 at 5 47 38 pm
Screen Shot 2021-07-20 at 12 52 48 pm

### CODE REVIEW

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+    private static final Logger LOGGER = Logger.getLogger(DeployedToEnvironmentNotifierStep.class.getName());
+
+    private final BitbucketDeploymentEnvironment environment;
+
+    @DataBoundConstructor
+    public DeployedToEnvironmentNotifierStep(String environmentKey, String environmentName,
+                                             @CheckForNull String environmentType,
+                                             @CheckForNull String environmentUrl) {
+        environment = new BitbucketDeploymentEnvironment.Builder(getOrGenerateEnvironmentKey(environmentKey),
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java]: you have a rather annoying bug here. Because you have `requireNonNull`  on the `key` and `name` it will throw an NPE, this exception will be displayed to the user with the full stack trace. Yes, they did the wrong thing, and no they should not get a stack trace thrown in their face because of it. Also because the message for the NPE is just "name" it is next to impossible to work out exactly what field it relates to. I am very well aware that some other plugins do throw exceptions in the users face on error. We need to do better than that.

@mhenschke-atl and I talked about this quickly, and essentially we have a number of bad options available to us. We both agree that current behaviour constitutes a bug, so we cannot ship with it. 

Options are
1. Provide a better message -> only marginally better, still a stack trace thrown at the user. Least preferred option
2. Accept that `null` is a thing, this is how the "trigger after build" works. This gives an acceptable user experience, and places the burden on you to validate that the Step is actually valid. Our Server config is done this way. This is not excellent but a decided improvement over the current situation
3. Do validation on save. This is more work, not quite as nice user experience as 2. but better technically (no `null` nor any invalid config being saved).

_So_ I propose we go with option 3. I think it strikes a balance between technical desires, and user experience. What you need to do is to override the `newInstance` in the descriptor like so:
```java
 @Override
        public Publisher newInstance(@Nullable StaplerRequest req, @Nonnull JSONObject formData) throws FormException {
            String envName = stripToNull(formData.getString("environmentName"));
            if (envName == null) {
                throw new FormException("Name is needed", "name");
            }
            return super.newInstance(req, formData);
        }

```
For bonus points (but totally optional) you can also handle the absence of the key, I don't think you need to though, as it is a form submit, so the key will only be missing when someone calls you directly and they can get a 500 or whatever if they send malformed requests, no user will get there by mistake.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+import java.util.logging.Logger;
+
+import static org.apache.commons.lang3.StringUtils.isBlank;
+import static org.apache.commons.lang3.StringUtils.stripToNull;
+
+public class DeployedToEnvironmentNotifierStep extends Notifier implements SimpleBuildStep {
+
+    private static final Logger LOGGER = Logger.getLogger(DeployedToEnvironmentNotifierStep.class.getName());
+
+    private final BitbucketDeploymentEnvironment environment;
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java:48]: I am not sure we really want to do it this way. The reason is that `BitbucketDeploymentEnvironment` now becomes API and _cannot_ be changed without an upgrade task to jobs. It will be serialised to disk as part of the job, subsequent reading will not call the constructor so it makes it hard to change things. 

Ultimately I think this is ok, but we really need to be aware of the upgrade pain we will have if we ever change the `BitbucketDeploymentEnvironment`

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            BitbucketSCMRepository bitbucketSCMRepo = revisionAction.getBitbucketSCMRepo();
+            String revisionSha = revisionAction.getRevisionSha1();
+            descriptor().deploymentPoster.postDeployment(bitbucketSCMRepo.getServerId(),
+                    bitbucketSCMRepo.getProjectKey(), bitbucketSCMRepo.getRepositorySlug(), revisionSha, deployment,
+                    run, listener);
+        } catch (RuntimeException e) {
+            // This shouldn't happen because deploymentPoster.postDeployment doesn't throw anything. But just in case,
+            // we don't want to throw anything and potentially stop other steps from being executed
+            String errorMsg =
+                    String.format("An exception occurred when trying to post the deployment to Bitbucket Server: %s", e.getMessage());
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java:108]:
```suggestion
                    String.format("An error occurred when trying to post the deployment to Bitbucket Server: %s", e.getMessage());
```

Our users may not be really familiar with `Exceptions` I guess most people are, but `error` conveys the same thing and being  less language specific.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Singleton
+public class DeploymentPosterImpl implements DeploymentPoster {
+
+    protected static final Logger LOGGER = Logger.getLogger(DeploymentPosterImpl.class.getName());
+
+    @Inject
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java:28]: I must say I would _very_ much prefer constructor injection instead. Makes it easier to breakpoint, log (during local development), and test (which is another thing there is no test for this class and there is enough logic below that I think it warrants a test)

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    private BitbucketPluginConfiguration pluginConfiguration;
+
+    @Override
+    public void postDeployment(String bbsServerId, String projectKey, String repositorySlug, String revisionSha,
+                               BitbucketDeployment deployment, Run<?, ?> run,
+                               TaskListener taskListener) {
+        Optional<BitbucketServerConfiguration> maybeServer = pluginConfiguration.getServerById(bbsServerId);
+        if (!maybeServer.isPresent()) {
+            taskListener.error(String.format("Could not send deployment notification to Bitbucket Server: Unknown serverId %s", bbsServerId));
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java:41]: I would statically import `format` so you don't have to drag the `String` qualifier in every time

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        }
+
+        taskListener.getLogger().printf("Sending notification of %s to %s on commit %s%n",
+                deployment.getState().name(), server.getServerName(), revisionSha);
+        try {
+            clientFactory.getProjectClient(projectKey)
+                    .getRepositoryClient(repositorySlug)
+                    .getDeploymentClient(revisionSha)
+                    .post(deployment);
+            taskListener.getLogger().printf("Successfully sent notification of %s deployment to %s on commit %s%n",
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java:61]: I know IntelliJ suggests this for you, but I find it harder to read when you alternate between `printf` and `String. format`. My preference would be to pick one and stick with it. It is a personal preference though, so make your own call on it.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+                    deployment.getState().name(), server.getServerName(), revisionSha);
+        } catch (AuthorizationException e) {
+            taskListener.error(String.format("Failed to send notification of deployment to %s due to an authorization error: %s",
+                    server.getServerName(), e.getMessage()));
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java:65]: `e.getMessage` is always  "Provided credentials cannot access the resource" So this will read
"Failed to send notification of deployment to {serverName} due to an authorization error: Provided credentials cannot access the resource" it feels long and clumsy for saying "Couldn't access the resource, check your password". Maybe we can make it better? what should I do if this happens in my build? what credentials are broken?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironmentType.java [hunk 8]
```diff
@@ -1,14 +1,48 @@
...
+    BitbucketDeploymentEnvironmentType(String displayName, int weight) {
+        this.displayName = displayName;
+        this.weight = weight;
+    }
+
+    public static Optional<BitbucketDeploymentEnvironmentType> fromName(@CheckForNull String name) {
+        if (StringUtils.isBlank(name)) {
+            return empty();
+        }
+        return Arrays.stream(values())
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironmentType.java:36]: Is this better than 
```java
 try {
            return Optional.of(valueOf(name.toUpperCase(Locale.ROOT)));
        } catch(IllegalArgumentException e) {
            return empty();
        }
```
?

**Review by dkjellin — CHANGES_REQUESTED** [review]
Partial review, I'll finish but there is already stuff to fix so I'll post it now

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+        return UUID.randomUUID().toString();
+    }
+
+    private static BitbucketDeploymentEnvironment validateAndGetEnvironment(String environmentKey,
+                                                                            String environmentName,
+                                                                            @CheckForNull String environmentType,
+                                                                            @CheckForNull String environmentUrl) {
+        FormValidation environmentNameValidation = validateEnvironmentName(environmentName);
+        if (!environmentNameValidation.equals(FORM_VALIDATION_OK)) {
+            throw new IllegalArgumentException(environmentNameValidation.getMessage());
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java:128]: This still suffers from the same problem (but with a much better message), I mentioned in my other comment: https://github.com/jenkinsci/atlassian-bitbucket-server-integration-plugin/pull/265#discussion_r673665818

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java (hunk 1 above)
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java]: Sorry, I changed this to 'option 1' before I saw your comment about it. Will go with Option 3.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java (hunk 4 above)
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java:28]: I said in the PR description (and mentioned it in stand-up this morning) that the unit tests are coming. I opened the PR for early feedback.

I don't understand why it makes it easier to test though, since `@InjectMocks` injects them just fine without it

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+                                             @CheckForNull String environmentType,
+                                             @CheckForNull String environmentUrl) {
+        environment = validateAndGetEnvironment(environmentKey, environmentName, environmentType, environmentUrl);
+    }
+
+    public DescriptorImpl descriptor() {
+        return Jenkins.get().getDescriptorByType(DescriptorImpl.class);
+    }
+
+    public BitbucketDeploymentEnvironment getEnvironment() {
```
Comment (ysun-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java:61]: are the following methods only used for tests? Or are they also going to be used by other classes in the future?
I feel like we can just keep `getEnvironment` since the other methods are accessors of the `environment`

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java (hunk 7 above)
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java:65]: We don't have a lot of visibility into which credentials are broken, as we get them from:
```java
GlobalCredentialsProvider globalCredentialsProvider = server.getGlobalCredentialsProvider(run.getParent());
Credentials globalAdminCredentials = globalCredentialsProvider.getGlobalAdminCredentials().orElse(null);
BitbucketCredentials credentials = jenkinsToBitbucketCredentials.toBitbucketCredentials(globalAdminCredentials);
```
Which means we may be making an anonymous call if there are no global credentials for the run. Additionally, `BitbucketCredentials` don't have any kind of id/name or anything else that we can use to identify it and neither do `Credentials`.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java (hunk 10 above)
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java:61]: The other methods are actually used by Jenkins to populate the fields on the page when you edit the job

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironmentType.java (hunk 8 above)
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironmentType.java:36]: I believe so, as exceptions to control flow are typically considered a code smell

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java (hunk 4 above)
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java:31]: >I said in the PR description (and mentioned it in stand-up this morning) that the unit tests are coming.

You totally did, I forgot it when I was here, sorry

>but I don't understand why a constructor makes it easier to test though since @InjectMocks injects them just fine without it

You are right, but you must now always either use the CI framework or `@injectMocs` to get hold of an instance. If you want to inject one custom class and the rest mocks, you have to use a `@spy`. If you want to quickly test something (not nervelessly a unit test, but a manual test) it is easier with constructor injection. Plus I see no real upside to having property injection, granted you don't have the constructor but that is not a huge benefit.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java (hunk 7 above)
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java:72]: That says exactly which credentials it is, it is the "credentials associated with the Bitbucket server instance. Please contact your admin to update the credentials"
Because you can only have one token per BbS instance and we always use that one to the exclusion of all others you know exactly which one it is and can provide guidance on how to fix it.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java (hunk 2 above)
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java]: Updated to have each text box match to a field (enviornmentName, environmentKey, etc)

**Comment by dkjellin** [comment]
I had so many multi-line suggestions that I put it all in a PR: https://github.com/jenkinsci/atlassian-bitbucket-server-integration-plugin/pull/266 take a look and decide what you think. We can talk more about it if you want.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java (hunk 4 above)
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java:30]: This still holds, I still very much prefer constructor injection over property injection.

**Review by dkjellin — APPROVED** [review]
I think this looks pretty good now.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+                    format("An error occurred when trying to post the deployment to Bitbucket Server: %s", e.getMessage());
+            listener.error(errorMsg);
+            LOGGER.info(errorMsg);
+            LOGGER.log(Level.FINE, "Stacktrace from deployment post failure", e);
+        }
+    }
+
+    private BitbucketDeploymentEnvironment getEnvironment(Run<?, ?> run, TaskListener listener) {
+        return new BitbucketDeploymentEnvironment(environmentKey,
+                getOrGenerateEnvironmentName(environmentName, run, listener),
```
[the remark this answers is not in the document]
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java:122]: The `environmentKey` needs to be generated once only, and then saved. It important that it doesn't change between runs. So it is done when the constructor is called.
When there is no environment name provided, we try and figure out something 'nice' based on the other fields. We use the environment type if that's provided. But if not, then we use the name of the run. I can fall back to the environment type in the constructor, but there's no way of knowing run information until the step is actually called

File: src/main/resources/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep/config.properties [hunk 12]
```diff
@@ -0,0 +1,9 @@
+bitbucket.deployments.env.description=Send notification of the deployment to Bitbucket Server. This requires the the job to have Bitbucket Server as the Source Code Management tool. The deployment information will be shown on commits and pull requests that are part of this deployment but are not part of a previously successful deployment to this environment.
+bitbucket.deployments.env.key.title=Environment Key
+bitbucket.deployments.env.key.help=This is a unique identifier for the environment in Bitbucket Server. You can name it something readable like MY-ENV, or you can leave it blank and have it auto-generated by the plugin.
```
Comment (EliSalter-atl) [inline src/main/resources/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep/config.properties:3]: /nit alphabetical order

File: src/main/resources/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep/help-environmentKey.html [hunk 13]
```diff
@@ -0,0 +1,6 @@
+<div>
```
Comment (EliSalter-atl) [inline src/main/resources/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep/help-environmentKey.html:1]: is there away to use the strings from config.properties here? so that we're not doubling up

**Review by EliSalter-atl — APPROVED** [review]
Looks good :)

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+    private final BitbucketDeploymentEnvironmentType environmentType;
+    private final String environmentUrl;
+
+    @DataBoundConstructor
+    public DeployedToEnvironmentNotifierStep(@CheckForNull String environmentKey,
+                                             @CheckForNull String environmentName,
+                                             @CheckForNull String environmentType,
+                                             @CheckForNull String environmentUrl) {
+        this.environmentKey = getOrGenerateEnvironmentKey(environmentKey);
+        this.environmentName = stripToNull(environmentName);
```
[the remark this answers is not in the document]
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java:64]: If it's blank we try and choose a sensible default.
See `getOrGenerateEnvironmentName`

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+    }
+
+    private String getOrGenerateEnvironmentKey(@CheckForNull String environmentKey) {
+        if (!isBlank(environmentKey)) {
+            return environmentKey;
+        }
+        return UUID.randomUUID().toString();
+    }
+
+    private String getOrGenerateEnvironmentName(@CheckForNull String environmentName, Run<?, ?> run,
```
[the remark this answers is not in the document]
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java:185]: As I mentioned in my [reply to Daniel's comment](https://github.com/jenkinsci/atlassian-bitbucket-server-integration-plugin/pull/265#discussion_r702551189), the generating the environment name is a little different to other methods we call in the constructor because it uses information from the `run`.
`getEnvironment` does not call to `getOrGenerateEnvironmentName`, so I'm not sure what 'each time' is that you're referring to

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+                                                TaskListener listener) {
+        if (!isBlank(environmentName)) {
+            return environmentName;
+        }
+        String generatedEnvironmentName;
+        if (environmentType != null) {
+            // Default the to the environment type display name if there is a configured environment type
+            generatedEnvironmentName = environmentType.getDisplayName();
```
[the remark this answers is not in the document]
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java:193]: Ideally they would provide the environment name in the configuration. It is a 'required' field in that form validation will fail if they don't provide it (see `doCheckEnvironmentName`). However, we don't want to block users from saving the job so we allow the invalid configuration, choose a smart default, and print out to the job's logs a message that their deployment step is misconfigured (see the call to `listener.getLogger()` a few lines below

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+        if (!maybeServer.isPresent()) {
+            taskListener.error(format("Could not send deployment notification to Bitbucket Server: Unknown serverId %s", bbsServerId));
+            return;
+        }
+
+        BitbucketServerConfiguration server = maybeServer.get();
+        Credentials globalAdminCredentials = server.getGlobalCredentialsProvider(run.getParent())
+                .getGlobalAdminCredentials()
+                .orElse(null);
+        BitbucketCredentials credentials = jenkinsToBitbucketCredentials.toBitbucketCredentials(globalAdminCredentials);
```
[the remark this answers is not in the document]
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java:57]: What would we do if it was `ANONYMOUS_CREDENTIALS`? From what I see in other places (e.g. `BuildStatusPoster`) we just attempt the request using whatever credentials we have.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironmentType.java [hunk 18]
```diff
@@ -1,14 +1,54 @@
...
+
+    private final String displayName;
+    private final int weight;
+
+    BitbucketDeploymentEnvironmentType(String displayName, int weight) {
+        this.displayName = displayName;
+        this.weight = weight;
+    }
+
+    public static Optional<BitbucketDeploymentEnvironmentType> fromName(@CheckForNull String name) {
```
[the remark this answers is not in the document]
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironmentType.java:40]: This isn't `fromDisplayName`. Passing through the display name will return `Optional.empty()`. This will only work if you pass through the enum name.

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImplTest.java [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+    @Mock
+    private PrintStream printStream;
+    @Mock
+    private BitbucketServerConfiguration server;
+
+    @Before
+    public void setup() {
+        FreeStyleProject parent = mock(FreeStyleProject.class);
+        when(run.getParent()).thenReturn(parent);
+        when(globalCredentialsProvider.getGlobalAdminCredentials()).thenReturn(of(globalAdminCredentials));
```
Comment (ysun-atl) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImplTest.java:82]: I think it would also be good to have a test for null credentials (or anonymous_credentials if we're going down that path)

**Review by ysun-atl — APPROVED** [review]

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+        BitbucketClientFactory clientFactory =
+                bitbucketClientFactoryProvider.getClient(server.getBaseUrl(), credentials);
```
[the remark this answers is not in the document]
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java:59]: Credentials won't be null because `jenkinsToBitbucketCredentials.toBitbucketCredentials` is not marked as `@CheckForNull` which means we can rely on it being non-null

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (EliSalter-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java:185]: I thought `getEnvironment` does. call to `getOrGenerateEnvironmentName` at [line 148](https://github.com/jenkinsci/atlassian-bitbucket-server-integration-plugin/pull/265/files/c664303044295b48c85fddc416cce8770205c5f6#diff-5aa1fc6c9c93213f91766e67821cdd817799b361038f930454aaca7c1dda9b5cR148). But maybe I'm understanding things wrong, so long as it's something we've considered

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (EliSalter-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java:57]: I was thinking that we would return and not attempt the request since we know we need credentials to post deployments and just add something to Jenkins logs. (or is it useful to have this info in BbS logs too)

**Review by dkjellin — APPROVED** [review]
